[smartthings] Initial contribution

BobRak opened the pull request:
This replaces pull request #5013 from Feb 28, 2019.  I made all of the changes requested there. And, refactored to code to match the layout used after the new build process was created. The long delay was due it: a) Some of the requested changes I could not get to work. Specifically making the discovery process into a component. b) getting a stable Eclipse development environment working. c) Personally I moved.

This is my initial contribution of a binding for the Samsung Smartthings hub. It has been in development since Jan 2017. It has been available in my repo for more than two years now and has been tested by a number of users. Smartthings defines a set of "capabilities" that every device must conform to. This binding supports many of the Smartthings capability. Many have been tested by me or others who have downloaded the binding. But, not all capabilities have been tested. And, some of the capabilities that Samsung added in the past year have not been implemented because they mostly involved Samsung appliances that I do not own.

Signed-off-by: Bob Raker rjraker@gmail.com (github: bobrak)

BobRak commented:
**Question about addition files required to use this binding**.
This binding requires that 2 additional files are installed into the smartthings hub. In my previous PR I asked where those files should be located and was told to create a **contrib** directory and place them there. Installation instructions are included in the `SmartthingsInstallation.md` file which is referenced in the readme.md file. 

Does the user need to go to github for those files or will they be located somewhere else?

fwolter commented on bundles/org.openhab.binding.smartthings/SmartthingsInstallation.md:
```diff
@@ -0,0 +1,79 @@
+# Installation of Smartthings code
```
I'm not sure if the location of the additional readme files, at the same level as the Readme.md is correct. @cpmeister Would `doc/` fit better?

cpmeister replied:
I believe that would be the correct place, yes.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+        super(thing);
+        this.smartthingsHandlerFactory = smartthingsHandlerFactory;
+        smartthingsName = ""; // Initialize here so it can be NonNull but it should always get a value in initialize()
+        config = getThing().getConfiguration().as(SmartthingsThingConfig.class);
```
configuration moved

cpmeister replied:
I think you forgot to remove this.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+     * Handle an update to the configuration
+     */
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
```
I removed it

BobRak commented on bundles/pom.xml:
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
When I look in my bundles/pom.xml it is identical to the recommended change. Maybe I'm missing something.

BobRak commented on .project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I have removed .project, bom/.project, bundles/.project,  I think these files were created by Eclipse when I setup the openHAB development environment. If that is true then why doesn't .gitignore include .project so they don't accidently get checked. Ok, am I missing something.

BobRak commented on bom/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
done

BobRak commented on bundles/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
done

BobRak commented on features/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
removed

fwolter commented:
> It never occurred to me that I should support more than on bridge.

From the user's perspective, it's possible to add multiple bridges. So, the user would expect, this is working. If I see correctly, there is not an error message generated when adding a second bridge, but the binding will behave unexpectedly. I think this is bad usability. You should at least document this behavior in the Readme.md.

fwolter commented:
Did you push all of your changes? The `.project` and pom.xml files, not belonging to your PR are still there.

BobRak commented:
No, I haven't pushed my changes yet. Probably a few days to a week. This
next week is going to be busy for me.  I still struggle with certain
aspects of git. It seems that clicking on the Resolve conversation button
updates my fork. My preference is to make all the changes and then I can do
a push.

On Sun, Jun 14, 2020 at 12:31 PM Fabian Wolter <notifications@github.com>
wrote:

> Did you push all of your changes? The .project and pom.xml files, not
> belonging to your PR are still there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab-addons/pull/7550#issuecomment-643811415>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHNNBA2Z77642MYTAF47PTRWUQRBANCNFSM4MZDQNSA>
> .
>

BobRak commented on bundles/org.openhab.binding.smartthings/SmartthingsInstallation.md:
```diff
@@ -0,0 +1,79 @@
+4. Verify your hub is listed.
+
+### Copying Smartthings files
+The files are located in the Git repo in the **contrib** directory structure.
+This is located in the openHAB Addons directory which is located [here](https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.smartthings/contrib).
```
Updated

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,235 @@
+        logger.debug("SmartthingsHandlerFactory is now processing ThingTypeUID {}", thingTypeUID.getAsString());
+
+        if (thingTypeUID.equals(THING_TYPE_SMARTTHINGS)) {
+            bridgeHandler = new SmartthingsBridgeHandler((Bridge) thing, this, bundleContext);
```
I have added code to only allow one bridge to be configured. And, I have added a statement in the README.md file stating only one bridge is allowed. The Smartthings hub has to be connected to a Location in a Samsung account. Only one hub is allowed per location.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,235 @@
+                result = gson.fromJson(responseStr, result.getClass());
+            }
+        } else if (status == 202) {
+            logger.info(
```
I removed the code related to status == 200 as it will never happen. 202 is the normal response code. The actual response message is sent by Smartthings as a separate message.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,235 @@
+                    "Sent message \"{}\" with path \"{}\" to the Smartthings hub, recieved HTTP status {} (This is the normal code from Smartthings)",
+                    data, path, status);
+        } else {
+            logger.info("Sent message \"{}\" with path \"{}\" to the Smartthings hub, recieved HTTP status {}", data,
```
I have never seen any result other that 202 but I expect a 500 could be returned if the hub was having a problem. I change the log level to warn. I think this is adequate.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    private @Nullable SmartthingsThingHandler findHandler(SmartthingsStateData stateData) {
+        for (SmartthingsThingHandler handler : thingHandlers) {
```
I made this method synchronized.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,235 @@
+        this.httpClient = httpClientFactory.getCommonHttpClient();
+    }
+
+    protected void unsetHttpClientFactory() {
```
Changed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,196 @@
+    private Logger logger = LoggerFactory.getLogger(SmartthingsServlet.class);
+    private @Nullable HttpService httpService;
+    private @Nullable EventAdmin eventAdmin;
+    private @Nullable Gson gson;
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,196 @@
+        this.eventAdmin = null;
+    }
+
+    // @Reference
+    protected void setGson(Gson gson) {
+        this.gson = gson;
+    }
+
+    protected void unsetGson() {
+        this.gson = null;
+    }
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,196 @@
+            logger.trace("Smartthing servlet processing \"error\" request. data: {}", sb);
+            Map<String, Object> map = new HashMap<String, Object>();
+            map = gson.fromJson(sb.toString(), map.getClass());
+            StringBuffer msg = new StringBuffer("Error message from Smartthings: ");
+            msg.append(map.get("message"));
+            logger.warn("{}", msg);
```
Building an error message to show the user. I am guessing you think it should be done differently?

fwolter replied:
You could build the log message simply inline. Would simplify code and improve performance a bit.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsColor100Converter.java:
```diff
@@ -0,0 +1,108 @@
+        // The easiest way to do this is to create a new HSBType with the hue component changed.
+        if (command instanceof HSBType) {
+            HSBType hsb = (HSBType) command;
+            double hue = (hsb.getHue().doubleValue() / 3.60) + 0.5; // add .5 to round
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsColorConverter.java:
```diff
@@ -0,0 +1,92 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Converter class for Door Control.
+ * This can't use the default because when closing the door the command that comes in is "closed" but "close" need to be
+ * sent to Smartthings
```
I've gone through all of the Color related converters and updated the javadoc to fit the class and better explain when they are used.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+    private Logger logger = LoggerFactory.getLogger(SmartthingsConverter.class);
+
+    protected String smartthingsName;
+    @Nullable
```
removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+    protected String thingTypeId;
+
+    SmartthingsConverter(Thing thing) {
+        smartthingsName = thing.getConfiguration().as(SmartthingsThingConfig.class).smartthingsName;
```
Not sure what you want here. It seems the other possibility is:
        smartthingsName = (String)thing.getConfiguration().get("smartthingsName");
Is that what you want or am I missing something?

fwolter replied:
Sorry, mixed it up with `BaseThingHandler.getConfigAs()`. Please ignore my comment.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+        } else if (command instanceof OnOffType) { // Need to surround with double quotes
+            value = surroundWithQuotes(command.toString().toLowerCase());
+        } else if (command instanceof OpenClosedType) { // Need to surround with double quotes
+            // OpenClosedType needs some tweeking. OpenClosedType.OPEN is fine but if the type is OpenClosedType.CLOSED
+            // need to send close, not closed.
```
removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+        } else if (command instanceof OpenClosedType) { // Need to surround with double quotes
+            // OpenClosedType needs some tweeking. OpenClosedType.OPEN is fine but if the type is OpenClosedType.CLOSED
+            // need to send close, not closed.
+            // String commandStr = (command.toString().equalsIgnoreCase("open")) ? "open" : "close";
+            // value = surroundWithQuotes(commandStr.toLowerCase());
```
removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+            value = surroundWithQuotes(command.toString().toLowerCase());
+        } else if (command instanceof PercentType) {
+            value = command.toString();
+        } else if (command instanceof PointType) { // Not really sure how to deal with this one and don't see a use for
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsConverter.java:
```diff
@@ -0,0 +1,218 @@
+        switch (acceptedChannelType) {
+            case "Color":
+                logger.warn(
+                        "Conversion of Color Contol-color is not currently supported. Need to provide support for message {}.",
```
Updated message

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsDefaultConverter.java:
```diff
@@ -0,0 +1,54 @@
+ * is used.
+ * If a channel specific converter is not found a convert based on the channel ID is used.
+ * If there is no convert found then this Default converter is used.
+ * Yes, it would be possible to change the SamrtthingsConverter class to not being abstract and implement these methods
+ * there. But, this makes it explicit that the default converter is being used.
```
Thank you for liking and agreeing with my approach here. The previous reviewer recommended combining SmartthingsConverter and SmarttingDefaultConverter. But, I preferred having them separate. That is the purpose for the comment.  Therefore I would prefer to leave it.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsHue100Converter.java:
```diff
@@ -0,0 +1,102 @@
+                    thingTypeId, smartthingsName, channelUid.getId(), value);
+
+        } else {
+            jsonMsg = defaultConvertToSmartthings(channelUid, command);
```
Yes, it should be logged. I added a log message in the SmartthingsConverter.defaultConvertToSmarttings method. All of the custom converters fall back to that method for things they can not convert.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+@Component(service = { DiscoveryService.class,
+        EventHandler.class }, immediate = true, configurationPid = "discovery.smartthings", property = "event.topics=org/openhab/binding/smartthings/discovery")
+public class SmartthingsDiscoveryService extends AbstractDiscoveryService implements EventHandler {
+    private static final int SEARCH_TIME = 30;
+    private static final int INITIAL_DELAY = 10; // Delay 10 sec to give time for bridge and things to be created
+    private static final int SCAN_INTERVAL = 600;
```
Changed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+        logger.debug("Initializing discovery service with default constructor.");
+    }
+
+    @Override
+    @Activate
+    protected void activate(@Nullable Map<String, @Nullable Object> config) {
+        super.activate(config);
+    }
```
I'm sorry, I don't understand your comment.

fwolter replied:
`super.activate()` should be invoked by the framework. So, you could remove the code above.

BobRak replied:
Removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+
+    @Reference
+    protected void setThingHandlerFactory(ThingHandlerFactory handlerFactory) {
+        logger.debug("Setting handlerFactory {}", handlerFactory);
```
removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+    @Reference
+    protected void setThingHandlerFactory(ThingHandlerFactory handlerFactory) {
+        logger.debug("Setting handlerFactory {}", handlerFactory);
+        smartthingsHandlerFactory = (SmartthingsHandlerFactory) handlerFactory;
```
Fixed. I didn't realize this method could be called from another binding. I actually had a user who downloaded the binding from my github report an exception here.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+        smartthingsHandlerFactory = (SmartthingsHandlerFactory) handlerFactory;
+    }
+
+    protected void unsetThingHandlerFactory(ThingHandlerFactory handlerFactory) {
+        logger.debug("Unsetting handlerFactory");
+        this.smartthingsHandlerFactory = null;
+    }
```
Done. Wish I had better understanding of the underlying OSGi framework.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+        logger.debug("SmartthingsDiscoveryService Starting background scan");
+        if (scanningJob == null || scanningJob.isCancelled()) {
+            logger.debug("Starting background scanning job");
+            if (scanningRunnable != null) {
```
Removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+        if (scanningJob == null || scanningJob.isCancelled()) {
+            logger.debug("Starting background scanning job");
+            if (scanningRunnable != null) {
+                this.scanningJob = scheduler.scheduleWithFixedDelay(this.scanningRunnable, INITIAL_DELAY, SCAN_INTERVAL,
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,255 @@
+     */
+    @Override
+    protected void stopBackgroundDiscovery() {
+        if (scanningJob != null && !scanningJob.isCancelled()) {
+            scanningJob.cancel(false);
+            scanningJob = null;
+        }
```
Updated.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/dto/SmartthingsDeviceData.java:
```diff
@@ -0,0 +1,58 @@
+        if (id == null) {
+            return "";
+        } else {
+            /* Note for code reviewer about the following retun statement:
+                Code analysis says There is no need for a @NonNull annotation because it is set as default. Only @Nullable should be used
+                But without it I get: Null type mismatch (type annotations): required '@NonNull String' but this expression has type '@Nullable String'
+                Better suggestions?
+            */
```
Done.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/dto/SmartthingsDiscoveryData.java:
```diff
@@ -0,0 +1,46 @@
+ */
+@NonNullByDefault
+public class SmartthingsDiscoveryData {
+    // private Logger logger = LoggerFactory.getLogger(SmartthingsDiscoveryData.class);
+    public long openHabStartTime;
+    public long hubTime;
+    // @Nullable
+    // public String @Nullable [] data;
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsBridgeHandler.java:
```diff
@@ -0,0 +1,133 @@
+import org.openhab.binding.smartthings.internal.SmartthingsBindingConstants;
+import org.openhab.binding.smartthings.internal.SmartthingsHandlerFactory;
+import org.osgi.framework.BundleContext;
+//import org.eclipse.equinox.event.
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsBridgeHandler.java:
```diff
@@ -0,0 +1,133 @@
+
+    private SmartthingsHandlerFactory smartthingsHandlerFactory;
+    private BundleContext bundleContext;
+    // private SmartthingsDiscoveryService smartthingsDiscoveryService;
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+                // Smartthings will not return a response to this message but will send it's response message
+                // which will get picked up by the SmartthingBridgeHandler.receivedPushMessage handler
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                logger.info("Attempt to send command to the Smartthings hub failed with exception", e);
```
Changed to warn and just log message

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+            @Nullable
+            String converterName = ch.getProperties().get("smartthings-converter"); // Will be null if no explicit
+                                                                                    // converter was specified
+            if (converterName == null || converterName.length() == 0) {
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+            }
+
+            // Try to get the converter
+            @Nullable
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,352 @@
+        } catch (ClassNotFoundException e) {
+            // Most of the time there is no channel specific converter, the default converter is all that is needed.
+            logger.trace("No Custom converter exists for {} ({})", converterName, converterClassName);
+        } catch (NoSuchMethodException e) {
+            logger.info("NoSuchMethodException occurred for {} ({}) {}", converterName, converterClassName,
+                    e.getMessage());
+        } catch (InvocationTargetException e) {
+            logger.info("InvocationTargetException occurred for {} ({}) {}", converterName, converterClassName,
+                    e.getMessage());
+        } catch (IllegalAccessException e) {
+            logger.info("IllegalAccessException occurred for {} ({}) {}", converterName, converterClassName,
+                    e.getMessage());
+        } catch (InstantiationException e) {
+            logger.info("InstantiationException occurred for {} ({}) {}", converterName, converterClassName,
+                    e.getMessage());
+        }
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Smartthings Binding</name>
+	<description>This is the binding for Smartthings.</description>
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+
+		<config-description>
+			<parameter name="smartthingsIp" type="text" required="true">
+				<label>IP address</label>
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+				<label>IP address</label>
+				<description>IP address of the Smartthings Hub</description>
+			</parameter>
+			<parameter name="smartthingsPort" type="integer" required="true">
```
Done. And, I added a default of 39500 which is the default port number specified by Smartthings

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+	<channel-type id="acceleration-channel">
+		<item-type>String</item-type>
+		<label>Smartthings Binding Channel</label>
+		<description>Sample channel for Smartthings Binding</description>
```
Fixed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+		<item-type>String</item-type>
+		<label>Door status</label>
+		<category>Door</category>
+		<!-- <state readOnly="true" pattern="%s"> <options> <option value="closed">Closed</option> <option value="closing">Closing</option> 
+			<option value="open">Open</option> <option value="opening">Opening</option> <option value="unknown">Unknown</option> </options> 
+			</state> -->
```
Restored. Don't remember why it was commented out

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+	</channel-type>
+
+	<channel-type id="windowShade-channel">
+		<item-type>String</item-type>
```
The Smartthings WindowShade definition doesn't map well to openHAB Rollershutter. Here is the Smartthing [WindowShade](https://docs.smartthings.com/en/latest/capabilities-reference.html#window-shade) definition. Unfortunately I don't have automated shades so I have n limited ability to test. But, my experience is that the String type gives the most flexibility.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+
+		<config-description>
+			<parameter name="smartthingsIp" type="text" required="true">
+				<label>IP address</label>
```
Fixed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,2284 @@
+				<label>acceleration</label>
+			</channel>
+		</channels>
+		<config-description>
+			<parameter name="smartthingsName" type="text" required="true">
+				<label>Smartthings Name</label>
+				<description>User assigned Smartthings device name</description>
+			</parameter>
+			<parameter name="smartthingsLocation" type="text">
+				<label>Smartthings Location</label>
+				<description>Where the device is located</description>
+			</parameter>
+		</config-description>
```
Done

fwolter commented:
I'm wondering why you have different repos for fetching and pulling. A typical configuration looks like this:
```
origin  https://github.com/fwolter/openhab-addons.git (fetch)
origin  https://github.com/fwolter/openhab-addons.git (push)
upstream        https://github.com/openhab/openhab-addons.git (fetch)
upstream        https://github.com/openhab/openhab-addons.git (push)
```

BobRak commented:
@fwolter 
OK, I changed my remotes to:
```
git remote -v
origin  https://github.com/BobRak/openhab-addons.git (fetch)
origin  https://github.com/BobRak/openhab-addons.git (push)
upstream        https://github.com/openhab/openhab-addons.git (fetch)
upstream        https://github.com/openhab/openhab-addons.git (push)
```
OK, now do I just need to do a `git pull origin` first? And then should `git push origin 1694-smartthings` work?

fwolter commented:
Yes, that should do the job. In case something goes wrong, it's good to have a backup of your local repo.

BobRak commented:
@fwolter 
Hi Fabian:

I was able to successfully pull the changes made in my origin repo down to my computer, resolve conflicts and then push. The steps I followed were:
```
git pull origin 1694-smartthings 
git config merge.tool vimdiff 
mergetool  
git commit -m "correction to README.md"
git push origin 1694-smartthings

```

That's the good new. Now the bad news. My repo is out of sync with upstream so I tried to resolve that with the following steps:
```
git fetch upstream
git branch –va
git checkout master
git merge upstream/master  
git checkout 1694-smartthings
git merge master   // 4000+ conflicts I’m stuck

```
Unfortunately this ended with >4000 conflicts. Here are a few lines:
```
CONFLICT (modify/delete): bundles/org.openhab.binding.airquality/src/main/resources/ESH-INF/binding/binding.xml deleted in master and modified in HEAD. Version HEAD of bundles/org.openhab.binding.airquality/src/main/resources/ESH-INF/binding/binding.xml left in tree.
CONFLICT (modify/delete): bundles/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/Attribute.java deleted in master and modified in HEAD. Version HEAD of bundles/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/Attribute.java left in tree.
CONFLICT (content): Merge conflict in bundles/create_openhab_binding_skeleton.cmd
Auto-merging bom/test-index/pom.xml
CONFLICT (content): Merge conflict in bom/test-index/pom.xml
Auto-merging bom/runtime-index/pom.xml
CONFLICT (content): Merge conflict in bom/runtime-index/pom.xml
Auto-merging bom/pom.xml
CONFLICT (content): Merge conflict in bom/pom.xml
Auto-merging bom/openhab-core-index/pom.xml
CONFLICT (content): Merge conflict in bom/openhab-core-index/pom.xml
Auto-merging bom/openhab-addons/pom.xml
CONFLICT (content): Merge conflict in bom/openhab-addons/pom.xml
Auto-merging README.md
CONFLICT (content): Merge conflict in README.md
Auto-merging CODEOWNERS
CONFLICT (content): Merge conflict in CODEOWNERS
Automatic merge failed; fix conflicts and then commit the result.
```
I know you are not supposed to be my git babysitter and I feel badly about asking you for so much git help. But, I'm stuck and don't know where to go from here. The only thing I can think of is to destroy my fork and start over. Do you have a recommendation? I did make a backup as you suggested so I have a place to start from.

BobRak commented:
@fwolter 
Me again...
I spent another 2 hours trying to figure out why I am getting so many code conflicts. Most of the conflicts are around the <version> in the pom file. I compared /pom from several repos and here is what I found:
| repo | /pom.xml version |
| --- | --- |
| 1694-smartthings (local repo) | 2.5.5-SNAPSHOT | 
| master (local repo) | 3.0.0-SNAPSHOT |
| origin/2.5.x | 2.5.5-SNAPSHOT | 
| origin/master | 3.0.0-SNAPSHOT | 
| upstream/2.5. | 2.5.7-SNAPSHOT |
| upstream/master| | 3.0.0-SNAPSHOT |

I'm not sure where things go out of wack. Can you suggest a path forward? Can you work off of what I have pushed? My only other idea would for me to delete my repo and then create a new fork.

Thanks again for your help.

fwolter commented:
Don't hesitate to ask for help. You can start over from your backup and finally push to this PR. This time use `git rebase` instead of `git merge` and use the branch `2.5.x` instead of `master` for rebasing.

BobRak commented:
@fwolter 
Fabian:

OK, I have the merge with upstream done and pushed   Thanks for all of your help.

Bob

BobRak commented on bundles/org.openhab.binding.smartthings/README.md:
```diff
@@ -0,0 +1,146 @@
+
+Discovery is not run automatically on startup. Therefore to run the discovery process perform the following:
+
+1. Start the PaperUI
+2. Click on **Configuration > Things**
+3. Click on **ADD THINGS** and select the Smartthings Binding.
+4. On the displayed screen select the Smartthings Hub.
+5. In the Configurations Parameters section enter the IP of Smartthings hub and enter the port number, which is probably 39500. Click on the check mark
+6. Click on **Inbox**
+7. At the bottom of the screen click on **SEARCH FOR THINGS**
+8. Select **Smartthings Binding**
+9. You should now see the Smartthings Things in the Inbox
+10. More information on using discovery is available in the [configuration tutorial](https://www.openhab.org/docs/configuration)
```
Done. Plus, discovery does run on startup now.

BobRak commented on bundles/org.openhab.binding.smartthings/README.md:
```diff
@@ -0,0 +1,146 @@
+
+## openHAB Configuration
+
+This binding is an openHAB 2 binding and uses the Bridge / Thing design with the Smartthings Hub being the Bridge and the controlled modules being the Things. The following definitions are specified in the .things file.
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/README.md:
```diff
@@ -0,0 +1,146 @@
+    String  SimulatedValve       "Simulated valve"                            { channel="smartthings:valve:Home:SimulatedValve:valve" }
+
+**Special note about Dimmers**
+There is a conceptual difference between how openHAB and Smartthings configures the dimmer and switch parts of a Dimmer. The Smartthings dimmer (capability name: switchLevel) is only able to accept a numeric value between 0 and 100 representing the brightness percentage. The openHAB dimmer is able to accept both the percentage and on/off. The openHAB PaperUI shows a dimmer with both a slider and  ...
```
I didn't realize that a dimmer can be configured as a switch in the sitemap. I removed this comment.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,200 @@
+        EventHandler.class }, immediate = true, configurationPid = "binding.smarthings", property = "event.topics=org/openhab/binding/smartthings/state")
+public class SmartthingsHandlerFactory extends BaseThingHandlerFactory implements ThingHandlerFactory, EventHandler {
+
+    private Logger logger = LoggerFactory.getLogger(SmartthingsHandlerFactory.class);
```
Done in all classes

BobRak commented on bundles/org.openhab.binding.smartthings/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>2.5.5-SNAPSHOT</version>
```
done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,200 @@
+
+    private Logger logger = LoggerFactory.getLogger(SmartthingsHandlerFactory.class);
+
+    private @NonNullByDefault({}) SmartthingsBridgeHandler bridgeHandler = null;
+    private @NonNullByDefault({}) ThingUID bridgeUID;
```
Changed to Nullable

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,200 @@
+    private @NonNullByDefault({}) SmartthingsBridgeHandler bridgeHandler = null;
+    private @NonNullByDefault({}) ThingUID bridgeUID;
+    private Gson gson;
+    private List<SmartthingsThingHandler> thingHandlers = new ArrayList<SmartthingsThingHandler>();
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java:
```diff
@@ -0,0 +1,200 @@
+
+    private @Nullable SmartthingsThingHandler findHandler(SmartthingsStateData stateData) {
+        for (SmartthingsThingHandler handler : thingHandlers) {
+            // There have been some reports of handler.getSmartthingsName() returning a null.
+            // Need to find out where null is coming from
+            if (handler.getSmartthingsName() != null) {
```
No I have not seen this exception in a very long time. I removed the null check. Certainly makes the code cleaner.

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,175 @@
+        try {
+            Dictionary<String, String> servletParams = new Hashtable<String, String>();
+            httpService.registerServlet(PATH, this, servletParams, httpService.createDefaultHttpContext());
+            logger.info("Started Smartthings servlet at {}", PATH);
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,175 @@
+    @Activate
+    protected void activate(Map<String, Object> config) {
+        if (httpService == null) {
+            logger.info("SmartthingsServlet.activate: httpService is unexpectedly null");
```
done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,224 @@
+        if (handlerFactory instanceof SmartthingsHandlerFactory) {
+            smartthingsHandlerFactory = (SmartthingsHandlerFactory) handlerFactory;
+        } else {
+            logger.debug("Ignoring setThingHandlerFactory for {}", handlerFactory.getClass().getName());
```
Done

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,224 @@
+        properties.put("smartthingsName", name);
+        properties.put("deviceId", deviceData.id);
+
+        DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(new ThingUID(uidStr)).withProperties(properties)
+                .withRepresentationProperty(deviceData.id).withBridge(bridgeUid).withLabel(name).build();
```
Changed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+    @Override
+    public void dispose() {
+        super.dispose();
+        logger.debug("Smartthings Handler disposed.");
```
Removed

BobRak commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,271 @@
+        // Create converters for each channel
+        for (Channel ch : thing.getChannels()) {
+            @Nullable
+            String converterName = ch.getProperties().get(smartthingsConverterName); // Will be null if no
+                                                                                     // explicit
+            // converter was specified
```
Changed

BobRak commented:
I knew this would take more than one pass. Thank you for turning things around so quickly. I am planning to leave home in mid July and not be back until late Sept. But, that could change if there is another lockdown in the US due to COVID. I'll be able to do a little remotely but it will be more difficult.

I'm still unsure what change you want on the pom.xml I just changed it to put the `<?xml..` and `<project` on the same line. I see other pom.xml files formatted this way.

fwolter commented:
Ok, great! Can you push your changes? Maybe we can get this merged before you leave.

The `project` tag should be wrapped after `...rg/2001/XMLSchema-instance"`. I.e. your PR should not contain any change in this line.

fwolter commented on bundles/pom.xml:
```diff
@@ -1,6 +1,4 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
These lines will still changed by your PR. You could copy those lines from the mainline file https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/pom.xml so that they won't appear in you PR.

BobRak replied:
done

BobRak commented:
@fwolter 
Fabian:
Is it too late for me to make a few changes mostly to the logging? Here is what I want to change:

1. I removed a DEBUG level log message that is printed when a new "state" message is received from Smartthings. I would like to restore that. It is really useful when trying to confirm that messages from Smartthings are being received and processed.
2. I left in a couple of DEBUG log messages when discovery data is received. I would change those to TRACE. Currently that is 90% of what is logged and it isn't that useful.
3. You asked if I could remove messages that contained timing data and I did that. However, I did not remove it from the code that is installed on the Smartthings hub. It doesn't do any harm but it isn't needed and may just be confusing to someone trying to make changes in the future.

Thanks,

Bob

fwolter commented:
Do it.

BobRak commented:
@fwolter 
Fabian: I've completed the changes I wanted to make.
Thanks,
Bob

fwolter commented:
LGTM. Can you sign off your latest commit?

BobRak commented:
@fwolter OK, signoff is complete.
Thanks,
Bob

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsHue100Converter.java:
```diff
@@ -0,0 +1,96 @@
+                double d = Double.parseDouble((String) deviceValue);
+                d *= 3.6;
+                return new DecimalType(d);
+            } else if (deviceValue instanceof Double) {
+                double d = ((Double) deviceValue).doubleValue();
+                d *= 3.6;
+                return new DecimalType(d);
+            } else if (deviceValue instanceof Long) {
+                double d = ((Long) deviceValue).longValue();
+                d *= 3.6;
+                return new DecimalType(d);
```
I think that just checking for `Number` instances would be a safer catchall 
```suggestion
            } else if (deviceValue instanceof Number) {
                double d = ((Number) deviceValue).doubleValue();
                d *= 3.6;
                return new DecimalType(d);
            }
```
You should also make a similar change in the `SmartthingsConverter` class.

BobRak replied:
OK to the change to SmartthingsHue100Converter

In SmartthingsConverter are you recommending changing:
```
            case "Number":
                if (deviceValue instanceof String) {
                    return new DecimalType(Double.parseDouble((String) deviceValue));
                } else if (deviceValue instanceof Double) {
                    return new DecimalType((Double) deviceValue);
                } else if (deviceValue instanceof Long) {
                    return new DecimalType((Long) deviceValue);
                } else {
                    logger.warn("Failed to convert Number {} with a value of {} from class {} to an appropriate type.",
                            deviceType, deviceValue, deviceValue.getClass().getName());
                    return UnDefType.UNDEF;
                }

```
To:
```
            case "Number":
                if (deviceValue instanceof String) {
                    return new DecimalType(Double.parseDouble((String) deviceValue));
                } else if (deviceValue instanceof Number) {
                    return new DecimalType((Number) deviceValue);
                } else if (deviceValue instanceof Long) {
                    return new DecimalType((Long) deviceValue);
                } else {
                    logger.warn("Failed to convert Number {} with a value of {} from class {} to an appropriate type.",
                            deviceType, deviceValue, deviceValue.getClass().getName());
                    return UnDefType.UNDEF;
                }

```
This doesn't compile because there is no constructor for **Number**. Or am I missing something?

cpmeister replied:
Since DecimalType doesn't have a Number constructor you would need to use one of the other ones.
```
            case "Number":
                if (deviceValue instanceof String) {
                    return new DecimalType(Double.parseDouble((String) deviceValue));
                } else if (deviceValue instanceof Long) {//we catch Long separately to maintain precision if necessary
                    return new DecimalType((Long) deviceValue);
                } else if (deviceValue instanceof Number) {
                    return new DecimalType(((Number) deviceValue).doubleValue());
                } else {
                    logger.warn("Failed to convert Number {} with a value of {} from class {} to an appropriate type.",
                            deviceType, deviceValue, deviceValue.getClass().getName());
                    return UnDefType.UNDEF;
                }
```
If you find it necessary, then you might want add another catch between the Long and Number catches for handling `BigDecimal` instances.

BobRak replied:
Sorry, I a little lost was tro what you want. I have changed the code to be:

```
        if ("Number".contentEquals(acceptedChannelType)) {
            if (deviceValue instanceof String) {
                double d = Double.parseDouble((String) deviceValue);
                d *= 3.6;
                return new DecimalType(d);
            } else if (deviceValue instanceof Long) {
                double d = ((Long) deviceValue).longValue();
                d *= 3.6;
                return new DecimalType(d);
            } else if (deviceValue instanceof BigDecimal) {
                double d = ((BigDecimal) deviceValue).doubleValue();
                d *= 3.6;
                return new DecimalType(d);
            } else if (deviceValue instanceof Number) {
                double d = ((Number) deviceValue).doubleValue();
                d *= 3.6;
                return new DecimalType(d);
            } else {
                logger.warn("Failed to convert Number {} with a value of {} from class {} to an appropriate type.",
                        deviceType, deviceValue, deviceValue.getClass().getName());
                return UnDefType.UNDEF;
            }
        } else {

```

If this isn't right please let me know and I'll change it again.

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsOpenCloseControlConverter.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public String convertToSmartthings(ChannelUID channelUid, Command command) {
+        String smartthingsValue = (command.toString().toLowerCase().equals("open")) ? "open" : "close";
+        smartthingsValue = (new StringBuilder()).append('"').append(smartthingsValue).append('"').toString();
```
if `surroundWithQuotes` was protected instead of private then you could reuse it here. 
```suggestion
        smartthingsValue = surroundWithQuotes(smartthingsValue);
```

BobRak replied:
Good idea. Done

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,217 @@
+
+    public Gson gson;
+
+    @Nullable
+    private SmartthingsHandlerFactory smartthingsHandlerFactory;
```
```suggestion
    private @Nullable SmartthingsHandlerFactory smartthingsHandlerFactory;
```

BobRak replied:
Done

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,217 @@
+    @Nullable
+    private SmartthingsHandlerFactory smartthingsHandlerFactory;
+
+    @Nullable
+    private ScheduledFuture<?> scanningJob;
```
```suggestion
    private @Nullable ScheduledFuture<?> scanningJob;
```

BobRak replied:
Done

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/discovery/SmartthingsDiscoveryService.java:
```diff
@@ -0,0 +1,217 @@
+
+    private final Logger logger = LoggerFactory.getLogger(SmartthingsDiscoveryService.class);
+
+    public Gson gson;
```
Can you make this private and final?

BobRak replied:
OK, done

cpmeister commented on bundles/org.openhab.binding.smartthings/pom.xml:
```diff
@@ -0,0 +1,16 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>2.5.7-SNAPSHOT</version>
```
```suggestion
    <version>2.5.9-SNAPSHOT</version>
```

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java:
```diff
@@ -0,0 +1,174 @@
+            case "error":
+                // This is an error message from smartthings
+                Map<String, Object> map = new HashMap<String, Object>();
+                map = gson.fromJson(s, map.getClass());
```
Does gson allow you to convert json into a `Map<String, Object>`? How would that even work, what class is it using for the `Object` values?

BobRak replied:
Yes this did work. The value (Object type) is a String. But, I changed it to <String, String>

cpmeister commented on bundles/org.openhab.binding.smartthings/contrib/smartthings/SmartApps/OpenHabAppV2.groovy:
```diff
@@ -0,0 +1,1052 @@
+
+// Massive lookup tree
+@Field CAPABILITY_MAP = [
+    "accelerationSensor": [
+        name: "Acceleration Sensor",
+        capability: "capability.accelerationSensor",
+        attributes: [
+            "acceleration"
+        ]
+    ],
```
Is this list generated or do you have to update by hand? You should look into finding a way to generate it automatically if possible.

BobRak replied:
This is based on the Smartthings capability and that is available as an xml file. In the future I can look at building an xslt that can transform it. It would also be nice to dynamically generate the **SUPPORTED_THING_TYPES_UIDS** set that is in SmartthingsBindingConstants.java. Are you aware of a way to do that?

cpmeister commented on bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/handler/SmartthingsThingHandler.java:
```diff
@@ -0,0 +1,270 @@
+        super(thing);
+        this.smartthingsHandlerFactory = smartthingsHandlerFactory;
+        smartthingsName = ""; // Initialize here so it can be NonNull but it should always get a value in initialize()
+        config = getThing().getConfiguration().as(SmartthingsThingConfig.class);
```
The configuration might not be valid at construction, so instead use a default config instance to make the nonnull annotation happy.
```suggestion
        config = new SmartthingsThingConfig();
```

BobRak replied:
Done

BobRak commented:
@cpmeister 
I have completed the changes your requested to the Smartthings binding. Hopefully I did what you wanted on the SmartthingsHue199Converter. In my testing it seems that the value coming in is always a string but I guess there are cases where that isn't true.

I read in the openHAB community email that 2.5.9 will be frozen on 9/20 which was news to me. And, that it will be the last 2.x release. I hope this can make that release as I am not looking forward to a openHAB 3 conversion but maybe it won't be that bad. I suffered through the new build process conversion and had lots of trouble getting the new build process working.

kaikreuzer reviewed (approved):
Hey @BobRak, that has been a long journey for you! It is awesome to see that smartthings support is finally there for openHAB 🎉.
After so many years of work, I also think we should get it into 2.5.9 now! Anticipating @cpmeister's approval, let's merge! (and if there are still some changes requested, let's move that to a follow-up PR).

But no worries: Moving development to openHAB 3 will be fairly easy. We have scripts that can port existing add-on code and the IDE setup is also fully automatic - so nothing to worry about!

BobRak commented:
@kaikreuzer Kai: Thanks for the kind words. Yes, it has been a long journey. I do think Smartthings is an important binding for openHAB and I am happy I could contribute it.  

Thanks to @cpmeister for his code review and help in making it better

And, a big THANK YOU to @fwolter for his code review and git support.  I am finally starting to get git.
